File: client/src/lib/poker-logic/player-actions.ts
import { logEvent } from "../analytics/analytics";
import type { HandState } from "./model";

export function playerCheck(seat: number, pokerState: HandState): HandState {
  if (isActionIsOutOfTurn(seat, pokerState)) return pokerState;
  if (pokerState.currentAction.minRaise > pokerState.seats[seat].currentRaise)
    return pokerState;

  logEvent("action-performed", {
    action: "check",
    seat,
    pokerState,
  });

  pokerState.seats[seat].lastAction = "Check";
  return pokerState;
}

export function playerFold(seat: number, pokerState: HandState): HandState {
  if (isActionIsOutOfTurn(seat, pokerState)) return pokerState;
  logEvent("action-performed", {
    action: "fold",
    seat,
    pokerState,
  });
  pokerState.seats[seat].lastAction = "Fold";
  pokerState.seats[seat].out = true;
  return pokerState;
}

export function playerCall(seat: number, pokerState: HandState): HandState {
  if (isActionIsOutOfTurn(seat, pokerState)) return pokerState;
  logEvent("action-performed", {
    action: "call",
    seat,
    pokerState,
  });
  pokerState.seats[seat].lastAction = "Call";
  const callAmount = Math.min(
    pokerState.seats[seat].stack,
    pokerState.currentAction.minRaise - pokerState.seats[seat].currentRaise
  );
  pokerState.seats[seat].stack -= callAmount;
  pokerState.pot += callAmount;
  return pokerState;
}

export function playerRaise(
  seat: number,
  pokerState: HandState,
  raiseAmount: number
): HandState {
  if (isActionIsOutOfTurn(seat, pokerState)) return pokerState;
  logEvent("action-performed", {
    action: "raise",
    seat,
    pokerState,
    raiseAmount,
  });
  const player = pokerState.seats[seat];
  pokerState.seats[seat].lastAction = "Raise";
  const totalRaiseAmount = raiseAmount;
  pokerState.seats[seat].stack -= totalRaiseAmount;
  pokerState.pot += raiseAmount;
  pokerState.currentAction.minRaise = raiseAmount + player.currentRaise;
  pokerState.seats[seat].currentRaise = pokerState.currentAction.minRaise;
  pokerState.currentAction.lastSeatToRaise = seat;
  return pokerState;
}

function isActionIsOutOfTurn(seat: number, pokerState: HandState): boolean {
  if (pokerState.currentAction.seatInTurn !== seat) {
    console.debug({
      message: "Seat attempted to act out of turn.",
      seat: seat,
      seatInTurn: pokerState.currentAction.seatInTurn,
    });
    return true;
  }
  return false;
}

File: pvp-server/src/logic/player-actions.go
package logic

import (
	"log"
	"math"
)

func PerformCheckFold(seat int, hand HandState) HandState {
	if isActionOutOfTurn(seat, hand) {
		return hand
	}
	if hand.CurrentAction.MinRaise > hand.Seats[seat].CurrentRaise {
		hand.Seats[seat].LastAction = "Fold"
		hand.Seats[seat].Out = true
	} else {
		hand.Seats[seat].LastAction = "Check"
	}

	return hand
}

func PlayerCall(seat int, hand HandState) HandState {
	log.Printf("Handling player call.")
	x := isActionOutOfTurn(seat, hand)

	if isActionOutOfTurn(seat, hand) {
		log.Printf("Wut? %v %v %v", seat, hand.CurrentAction.SeatInTurn, x)
		return hand
	}
	log.Printf("Handling player call also.")
	hand.Seats[seat].LastAction = "Call"
	callAmount := int(math.Min(float64(hand.Seats[seat].Stack), float64(hand.CurrentAction.MinRaise-hand.Seats[seat].CurrentRaise)))
	hand.Seats[seat].Stack -= callAmount
	log.Printf("Calling. callAmount[%v] pot[%v]\n", callAmount, hand.Pot)
	hand.Pot += callAmount
	return hand
}

func PlayerRaise(seat int, hand HandState, raiseAmount int) HandState {
	if isActionOutOfTurn(seat, hand) == true {
		return hand
	}
	player := hand.Seats[seat]
	player.LastAction = "Raise"
	player.Stack -= raiseAmount
	hand.Pot += raiseAmount
	hand.CurrentAction.MinRaise = raiseAmount + player.CurrentRaise
	player.CurrentRaise = raiseAmount + player.CurrentRaise
	hand.CurrentAction.LastSeatToRaise = seat
	hand.Seats[seat] = player
	return hand
}

func isActionOutOfTurn(seat int, hand HandState) bool {
	if hand.CurrentAction.SeatInTurn != seat {
		log.Println("Action performed out of turn.")
	}
	return hand.CurrentAction.SeatInTurn != seat
}
